### PR TITLE
Fix max iters issue and add tests

### LINKED ISSUE #1521: Possible issues with max_iters when loading/saving engine's state
Currently Engine implicitly assumes : 
https://github.com/pytorch/ignite/blob/e3ef192c94d0d793a9303bec915fb846aaa3161f/ignite/engine/engine.py#L122

but we recently introduced possibility to trigger a run with `max_iters`. If epoch length is unknown, saving and reloading engine's state probably wont work. We have to ensure that everything still works as expected.

@thescripted would you be interested in looking into this ?

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #1521

Depends on:
- https://github.com/pytorch/ignite/pull/3442
- https://github.com/pytorch/ignite/pull/3443
- https://github.com/pytorch/ignite/pull/3444
- https://github.com/pytorch/ignite/pull/3445

### CODE REVIEW

File: ignite/metrics/mean_average_precision.py [hunk 1]
```diff
@@ -226,7 +226,7 @@ def __init__(
             device=device,
             skip_unrolling=skip_unrolling,
         )
-        super(Metric, self).__init__(rec_thresholds=rec_thresholds, class_mean=class_mean)
+        _BaseAveragePrecision.__init__(self, rec_thresholds=rec_thresholds, class_mean=class_mean)
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply ignite/metrics/mean_average_precision.py]: The base classes do not have the same params, so we need to initialize them separately

File: ignite/metrics/vision/object_detection_average_precision_recall.py [hunk 2]
```diff
@@ -129,7 +129,8 @@ def box_iou(pred_boxes: torch.Tensor, gt_boxes: torch.Tensor, iscrowd: torch.Boo
             device=device,
             skip_unrolling=skip_unrolling,
         )
-        super(Metric, self).__init__(
+        _BaseAveragePrecision.__init__(
+            self,
```
[the remark this answers is not in the document]
Comment (goanpeca) [reply ignite/metrics/vision/object_detection_average_precision_recall.py]: The base classes do not have the same params, so we need to initialize them separately

File: .github/workflows/code-style-checks.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
+name: Code Style Checks
```
Comment (vfdev-5) [inline .github/workflows/code-style-checks.yml]: Let's split style check and GH workflow related changes to another PR, please

File: docs/source/conf.py [hunk 4]
```diff
@@ -12,6 +12,7 @@
 # add these directories to sys.path here. If the directory is relative to the
 # documentation root, use os.path.abspath to make it absolute, like shown here.
 #
+import doctest
```
Comment (vfdev-5) [inline docs/source/conf.py]: Let's put this into a separate PR, please

File: ignite/engine/engine.py [hunk 5]
```diff
@@ -1346,6 +1474,16 @@ def _run_once_on_dataset_legacy(self) -> float:
 
         return time.time() - start_time
 
+    def debug(self, enabled: bool = True) -> None:
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: Engine debug mode feature request is described here: https://github.com/pytorch/ignite/issues/1992
Let's revert this revert this change and tackle the feature later in another PR. Thanks!

File: ignite/handlers/clearml_logger.py [hunk 6]
```diff
@@ -118,6 +118,8 @@ class ClearMLLogger(BaseLogger):
 
     """
 
+    _task: Any  # Type annotation for ClearML Task
```
Comment (vfdev-5) [inline ignite/handlers/clearml_logger.py]: Let's put typo annotation fixes into another PR, maybe with code style GH workflow changes, please

File: ignite/handlers/wandb_logger.py [hunk 7]
```diff
@@ -26,7 +26,7 @@ class WandBLogger(BaseLogger):
     Args:
         args: Positional arguments accepted by `wandb.init`.
         kwargs: Keyword arguments accepted by `wandb.init`.
-            Please see `wandb.init <https://docs.wandb.ai/ref/python/init>`_ for documentation of possible parameters.
+            Please see `wandb.init <https://docs.wandb.ai/ref/python/sdk/functions/init/>`_ for documentation of possible parameters.
```
Comment (vfdev-5) [inline ignite/handlers/wandb_logger.py]: Let's put this into another PR

File: ignite/metrics/gan/fid.py [hunk 8]
```diff
@@ -37,7 +37,7 @@ def fid_score(
     diff = mu1 - mu2
 
     # Product might be almost singular
-    covmean, _ = scipy.linalg.sqrtm(sigma1.mm(sigma2), disp=False)
+    covmean, _ = scipy.linalg.sqrtm(sigma1.mm(sigma2).detach().cpu().numpy(), disp=False)
```
Comment (vfdev-5) [inline ignite/metrics/gan/fid.py]: This is a bit strange that we need to call detach and twice cpu. Sigma1 and sigma2 are already on cpu. Maybe, we should detach on the lines 34, 35:
```python
    mu1, mu2 = mu1.detach().cpu(), mu2.detach().cpu()
    sigma1, sigma2 = sigma1.detach().cpu(), sigma2.detach().cpu()
```

Let's put this change into another PR, please

File: ignite/metrics/vision/object_detection_average_precision_recall.py [hunk 9]
```diff
@@ -124,12 +124,16 @@ def box_iou(pred_boxes: torch.Tensor, gt_boxes: torch.Tensor, iscrowd: torch.Boo
         self._area_range = area_range
         self._max_detections_per_image_per_class = max_detections_per_image_per_class
 
-        super(ObjectDetectionAvgPrecisionRecall, self).__init__(
+        # Initialize Metric first
```
Comment (vfdev-5) [inline ignite/metrics/vision/object_detection_average_precision_recall.py]: Why we should do this?
I'm not totally following why we should do this change. If you can please explain it, would be great! Thanks!

File: ignite/metrics/mean_average_precision.py [hunk 10]
```diff
@@ -220,13 +220,16 @@ def __init__(
         .. versionadded:: 0.5.2
         """
 
-        super(MeanAveragePrecision, self).__init__(
+        # Initialize _BaseClassification first
```
Comment (vfdev-5) [inline ignite/metrics/mean_average_precision.py]: I'm not totally following why we should do this change. If you can please explain it, would be great! Thanks!
Can we do this in a separate PR?

File: mypy.ini [hunk 11]
```diff
@@ -22,7 +22,7 @@ warn_return_any = False
 ; results in too many false positives, therefore set to False
 warn_unreachable = False
 warn_unused_configs = True
-warn_unused_ignores = True
+warn_unused_ignores = False
```
Comment (vfdev-5) [inline mypy.ini]: Why we need this change?

**Review by vfdev-5 — COMMENTED** [review]
Thanks a lot for the PR @goanpeca !
I made the first pass and left comments essentially to split this PR into multiple ones and explain certain changes.
I'll check in more depth the max_iters related changes a bit later whether it works as expected.
Anyway, great job!

File: ignite/handlers/clearml_logger.py (hunk 6 above)
Comment (goanpeca) [reply ignite/handlers/clearml_logger.py]: https://github.com/pytorch/ignite/pull/3442

File: .github/workflows/code-style-checks.yml (hunk 3 above)
Comment (goanpeca) [reply .github/workflows/code-style-checks.yml]: https://github.com/pytorch/ignite/pull/3442

File: docs/source/conf.py (hunk 4 above)
Comment (goanpeca) [reply docs/source/conf.py]: https://github.com/pytorch/ignite/pull/3443

File: ignite/handlers/wandb_logger.py (hunk 7 above)
Comment (goanpeca) [reply ignite/handlers/wandb_logger.py]: https://github.com/pytorch/ignite/pull/3443

File: ignite/metrics/gan/fid.py (hunk 8 above)
Comment (goanpeca) [reply ignite/metrics/gan/fid.py]: https://github.com/pytorch/ignite/pull/3444

File: mypy.ini (hunk 11 above)
Comment (goanpeca) [reply mypy.ini]: I was getting mypy errors on CI that I could not reproduce locally! so the fixes for CI, were detected as false positives locally.

Not sure what to do about it @vfdev-5

File: ignite/metrics/vision/object_detection_average_precision_recall.py (hunk 9 above)
Comment (goanpeca) [reply ignite/metrics/vision/object_detection_average_precision_recall.py]: - MRO: ObjectDetectionAvgPrecisionRecall → Metric → 
  Serializable → _BaseAveragePrecision
  - `super(Metric, self)` goes to `Serializable`, not `_BaseAveragePrecision` so something needs to be fixed in the way the MRO is working!

File: ignite/metrics/mean_average_precision.py (hunk 10 above)
Comment (goanpeca) [reply ignite/metrics/mean_average_precision.py]: Similar issue. see [other comment](https://github.com/pytorch/ignite/pull/3439#discussion_r2323141723).

File: ignite/metrics/vision/object_detection_average_precision_recall.py (hunk 9 above)
Comment (vfdev-5) [reply ignite/metrics/vision/object_detection_average_precision_recall.py]: This is not related to this feature right?
Let's do this in a separte PR please

Comment (goanpeca) [reply ignite/metrics/vision/object_detection_average_precision_recall.py]: Ok, will split it :)

File: mypy.ini (hunk 11 above)
Comment (vfdev-5) [reply mypy.ini]: Let's put this into another PR

File: tests/ignite/conftest.py [hunk 12]
```diff
@@ -48,7 +48,7 @@ def pytest_configure(config):
     config.addinivalue_line("markers", "distributed: run distributed")
     config.addinivalue_line("markers", "multinode_distributed: distributed")
     config.addinivalue_line("markers", "tpu: run on tpu")
-    if config.option.treat_unrun_as_failed:
+    if hasattr(config.option, "treat_unrun_as_failed") and config.option.treat_unrun_as_failed:
```
Comment (vfdev-5) [inline tests/ignite/conftest.py]: Same here, let's put this into another PR

File: ignite/engine/engine.py [hunk 13]
```diff
@@ -691,14 +691,20 @@ def save_engine(_):
                 a dictionary containing engine's state
 
         """
-        keys: Tuple[str, ...] = self._state_dict_all_req_keys + (self._state_dict_one_of_opt_keys[0],)
+        keys: Tuple[str, ...] = self._state_dict_all_req_keys
+        keys += ("iteration",)
```
Comment (aaishwarymishra) [inline ignite/engine/engine.py]: what about ` "epoch"`

File: ignite/base/mixins.py [hunk 14]
```diff
@@ -1,11 +1,18 @@
...
+from typing import List, Tuple
 
 
 class Serializable:
-    _state_dict_all_req_keys: Tuple = ()
-    _state_dict_one_of_opt_keys: Tuple = ()
+    _state_dict_all_req_keys: Tuple[str, ...] = ()
+    _state_dict_one_of_opt_keys: Tuple[Tuple[str, ...], ...] = ((),)
+
+    def __init__(self) -> None:
```
Comment (TahaZahid05) [inline ignite/base/mixins.py]: Some subclasses like `EarlyStopping`, `Checkpoint`, `Metric` never call `super().__init__()`. I think a better approach would be to revert it so `_state_dict_user_keys` stays in Engine as not all subclasses need this.

File: ignite/engine/engine.py [hunk 15]
```diff
@@ -756,6 +792,59 @@ def _is_done(state: State) -> bool:
...
+            if (self.state.max_iters is not None) and max_iters < self.state.iteration:
+                raise ValueError(
+                    "Argument max_iters should be greater than or equal to the start "
+                    f"iteration defined in the state: {max_iters} vs {self.state.iteration}. "
+                    "Please, set engine.state.max_iters = None "
+                    "before calling engine.run() in order to restart the training from the beginning."
+                )
+            self.state.max_iters = max_iters
+
+    def _check_and_set_epoch_length(self, data: Optional[Iterable], epoch_length: Optional[int] = None) -> None:
```
Comment (TahaZahid05) [inline ignite/engine/engine.py:825]: This function is not used anywhere. Best to remove it.

File: ignite/engine/engine.py [hunk 16]
```diff
@@ -854,59 +943,98 @@ def switch_batch(engine):
...
+            )
+            or (
+                self._is_done(self.state)
+                and self._internal_run_generator is None
+                and (
+                    (max_epochs is not None and max_epochs == self.state.max_epochs)
+                    or (max_iters is not None and max_iters == self.state.max_iters)
+                )
+            )
+        )
```
Comment (TahaZahid05) [inline ignite/engine/engine.py:973]: We should discuss this more. Why are we creating a new state only if the same parameters passed and resuming otherwise? I personally think the master behavior of always restarting is much more intutive. Current proposed change can be misleading for the user. 

For instance, on master, `engine.run(data, max_epochs=10)` after completing 5 epochs would restart from 0 but this PR is proposing for it to continue from 5 and go to 10, while for `engine.run(data, max_epochs=5)` (same value as initial run), it would restart. This can be confusing

File: ignite/base/mixins.py [hunk 17]
```diff
@@ -1,11 +1,18 @@
...
+        self._state_dict_user_keys: List[str] = []
+
+    @property
+    def state_dict_user_keys(self) -> List:
+        return self._state_dict_user_keys
```
Comment (TahaZahid05) [inline ignite/base/mixins.py:15]: Some subclasses like `EarlyStopping`, `Checkpoint`, `Metric` never call `super().__init__()`. I think a better approach would be to revert it so `_state_dict_user_keys` stays in Engine as not all subclasses need this.

File: tests/ignite/engine/test_max_iters_fix.py [hunk 18]
```diff
@@ -0,0 +1,423 @@
...
+    sd = engine1.state_dict()
+    counter[0] = 0
+
+    # Second engine: load state and increase max_iters
+    engine2 = Engine(update_fn)
+    engine2.load_state_dict(sd)
+
+    # Directly set higher max_iters and run
+    engine2.state.max_iters = 25
+    engine2.run(data)
```
Comment (TahaZahid05) [inline tests/ignite/engine/test_max_iters_fix.py:415]: Why not do `engine2.run(data, max_iters=25)` instead? This way we can verify the actural resume path of `run()`

File: tests/ignite/engine/test_max_iters_fix.py [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+    """Test load_state_dict with max_iters."""
+    engine = Engine(lambda e, b: 1)
+
+    state_dict = {"iteration": 150, "max_iters": 250, "epoch_length": 100}
+
+    engine.load_state_dict(state_dict)
+    assert engine.state.iteration == 150
+    assert engine.state.max_iters == 250
+    assert engine.state.epoch_length == 100
+    assert engine.state.epoch == 1  # 150 // 100
```
Comment (TahaZahid05) [inline tests/ignite/engine/test_max_iters_fix.py:65]: Lets test `epoch` + `max_iters` as well

File: tests/ignite/engine/test_max_iters_fix.py
Comment (TahaZahid05) [inline tests/ignite/engine/test_max_iters_fix.py:1]: A Checkpoint test case would also be good since it is quite a common use so best to test it out.

**Review by TahaZahid05 — CHANGES_REQUESTED** [review]

**Comment by vfdev-5** [comment]
Closing in favor of https://github.com/pytorch/ignite/pull/3729
